### ISSUE TITLE
Flash

### ISSUE THREAD

**Issue by apaganobeleno** [gobuffalo/buffalo#139]
@markbates how do we deal with Flash when using Buffalo ? something like:

```go
//inside some action
Flash.Set("error", "Please check your data.")
```

would be great to have, thoughts ?

**Comment by markbates** [comment]
A flash object would be great. Would love to see a proposal on how it would be done. From handler down to template.

**Comment by apaganobeleno** [comment]
Great, i'll work on that propose and keep you posted, thanks!

**Comment by bketelsen** [comment]
yes!!

**Comment by bscott** [comment]
+1

**Comment by apaganobeleno** [comment]
@markbates i'll try to summarize my first shot at this one on this issue (i have it working right now) along with some questions i could have at this point.

We could do this by adding:

```go
Flash() *Flash
```

Where flash is a Flash struct thats defined something like:

```go
type Flash struct {
    context *buffalo.Session
}
```

And has `Get` and `Set` implementations inside it so that on the actions level we could do something like:

```go
c.Flash.Set("error", "Please enter valid credentials.")
c.Flash.Get("error")
```

And as for the View level on the action we could be doing something like:

```go

func ViewFlashAction(c buffalo.Context) error {
	val := c.Flash().Get("error")
	return c.Render(200, r.HTML("view_flash.html"))
}
```

With its corresponding view:

```html
{{error}}
```

Hope this make sense!

**Comment by apaganobeleno** [comment]
Oh the questions:
- Should we add the Flash object inside the Session ?
- Should we preload Flash variables so we don't have to call c.Flash ?
- Should we provide a Flash helper function that could be called from the view ?

**Comment by bscott** [comment]
@apaganobeleno looks great, I can't answer the questions above but I can't wait to try it, been really wanting this feature. Using c.Flash, feels right to me but not my call.

**Comment by markbates** [comment]
Ok, finally had a chance to look at this. :) Sorry for the delay.

I think adding `Flash()` to the context is the right place:

```go
c.Flash().Set("error", "Please enter valid credentials.")
c.Flash().Get("error")
```

In the view it should be something like this:

```handlebars
{{flash.error}}
{{flash.notice}}
etc...

{{#each flash as |k v|}}
<div class="{{k}}">{{v}}</div>
{{/each}}
```

Things to think about:

* There should be an `Add` function that will add a flash message to that key, so the messages should be probably be stored like `map[string][]string`.
* `Set` should override any messages previously stored.
* `Delete` should delete any messages with a given key
* `Clear` should wipe out all of the messages, regardless of key
* When a new Context is created the underlying `Flash` object should be created with empty values and set onto the page context so users can check the flash stuff in templates and it should work
* Flash messages need to be cleared once they are displayed. Not sure how to do this, perhaps have the store be of type `map[string]FlashMessage`. Perhaps clear them after a successful template render?? Not sure.
* messages should persist across a redirect

I think that about covers it! :) I'm really excited for this feature. I can't wait to see what you get done with this.

Thanks!

**Comment by bketelsen** [comment]
Error
Notice
Success

<-- my votes for levels.

**Comment by markbates** [comment]
Ok, I had some thoughts on how to clear the flash messages!

```handlebars
{{ flash "error" }}

{{#flashes }}
iterate through flashes
{{/flashes}}
```

When the `flash` or `flashes` helpers are called they clear out the messages! :)

**Comment by bketelsen** [comment]
like a defer() built into the iterator?

**Comment by markbates** [comment]
Yeah, something like that. As you display messages, get rid of them.

**Comment by apaganobeleno** [comment]
Awesome, i have the `Flash` part all-set, need to dig into the render to work on the View implementation, i'll push my WIP soon to a branch here, thanks @markbates @bketelsen @bscott for the support!

**Comment by bscott** [comment]
The suspense is killing me :)

**Comment by apaganobeleno** [comment]
lol sorry for that

**Comment by markbates** [comment]
Closed by #143 

For historical purposes, we landed on the following use case:

```go
// Create default implementation.
func PostsCreate(c buffalo.Context) error {
        // do some work
	c.Flash().Add("success", "Post was successfully created!")
	return c.Redirect(302, "/posts/%s", post.ID)
}
```

```handlebars
{{#each flash as |k messages|}}
  {{#each messages as |msg|}}
    <div class="alert alert-{{k}}" role="alert">
      <button type="button" class="close" data-dismiss="alert" aria-label="Close"><span aria-hidden="true">&times;</span></button>
      {{msg}}
    </div>
  {{/each}}
{{/each}}

// or

{{#each flash.errors as |k msg|}}
  <div class="alert alert-danger" role="alert">
    <button type="button" class="close" data-dismiss="alert" aria-label="Close"><span aria-hidden="true">&times;</span></button>
    {{msg}}
  </div>
{{/each}}

```
